test/Test/Database/LSMTree/Internal/Serialise/Class.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Test.Database.LSMTree.Internal.Serialise.Class (tests) where
+
+import           Data.ByteString (ByteString)
+import           Data.ByteString.Lazy (LazyByteString)
+import           Data.ByteString.Short (ShortByteString)
+import           Data.Primitive (ByteArray)
+import           Data.Word
+import           Database.LSMTree.Extras.Generators ()
+import           Database.LSMTree.Internal.Serialise
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Serialise.Class"
+    [ allProperties @Word64 "Word64"
+    , allProperties @ByteString "ByteString"
+    , allProperties @LazyByteString "LazyByteString"
+    , allProperties @ShortByteString "ShortByteString"
+    , allProperties @ByteArray "ByteArray"
+    ]
+
+allProperties :: forall a. (Ord a, Show a, Arbitrary a, SerialiseKey a, SerialiseValue a) => String -> TestTree
+allProperties name =
+    testGroup name
+      [ testProperty "prop_roundtripSerialiseKey" $
+          prop_roundtripSerialiseKey @a
+      , testProperty "prop_roundtripSerialiseValue" $
+          prop_roundtripSerialiseValue @a
+      , testProperty "prop_orderPreservationSerialiseKey" $
+          prop_orderPreservationSerialiseKey @a
+      , testProperty "prop_orderPreservationSerialiseValue" $
+          prop_orderPreservationSerialiseValue @a
+      ]
+
+prop_roundtripSerialiseKey :: (Eq k, Show k, SerialiseKey k) => k -> Property
+prop_roundtripSerialiseKey k =
+    deserialiseKey (serialiseKey k) === k
+
+prop_roundtripSerialiseValue :: (Eq k, Show k, SerialiseValue k) => k -> Property
+prop_roundtripSerialiseValue v =
+    deserialiseValue (serialiseValue v) === v
+
+prop_orderPreservationSerialiseKey :: (Ord k, SerialiseKey k) => k -> k -> Property
+prop_orderPreservationSerialiseKey x y =
+    compare x y === compare (serialiseKey x) (serialiseKey y)
+
+prop_orderPreservationSerialiseValue :: (Ord k, SerialiseValue k) => k -> k -> Property
```

Comment:
there isn't requirement for this, is there?

Reply:
The property is currently specified in the Haddock: https://github.com/IntersectMBO/lsm-tree/blob/9f7ac2896cba4563d4135307cb36ac7957daa58c/src/Database/LSMTree/Internal/Serialise/Class.hs#L37

But I also think there is no *real* requirement to have that property and and would be best to just get rid of it (option 3 I listed). Just not sure others agree, there might be some reason I'm missing.

Reply:
these are `SerialiseKey` haddocks.

Reply:
Yes, this property can be removed

Reply:
Yes, I shouldn't use it for values.